trin-execution/src/storage/evm_db.rs

```diff
@@ -109,150 +101,250 @@ impl EvmDB {
+                self.db.get(address_hash)?.map(|raw_account| {
+                    Decodable::decode(&mut raw_account.as_slice())
+                        .expect("valid account should be decoded")
+                });
+            stop_timer(timer);
+
+            let rocks_account = if wipe_storage && rocks_account.is_some() {
+                let rocks_account = rocks_account.expect("We already checked that it is some");
+                self.delete_account_storage(address_hash, rocks_account, false)?
             } else {
-                RocksAccountState::Touched
+                rocks_account
             };
 
-            for (key, value) in account
-                .storage
-                .into_iter()
-                .filter(|(_, value)| value.is_changed())
-            {
-                if value.present_value() > U256::ZERO {
-                    let _ = trie.insert(
-                        keccak256(B256::from(key)).as_ref(),
-                        &alloy_rlp::encode(value.present_value()),
-                    );
-                } else {
-                    let _ = trie.remove(keccak256(B256::from(key)).as_ref());
-                }
+            if !storage.is_empty() {
+                self.commit_storage_changes(address_hash, rocks_account, storage)?;
             }
+            stop_timer(plain_state_storage_timer);
+        }
+        Ok(())
+    }
 
-            // update trie
-            let RootWithTrieDiff {
-                root: storage_root,
-                trie_diff,
-            } = trie
-                .root_hash_with_changed_nodes()
-                .expect("Getting the root hash should never fail");
-
-            if self.config.cache_contract_storage_changes {
-                let account_storage_cache = self.storage_cache.entry(address_hash).or_default();
-                for key in trie_diff.keys() {
-                    account_storage_cache.insert(*key);
-                }
-            }
+    pub fn commit_bundle(&mut self, bundle_state: BundleState) -> anyhow::Result<()> {
+        // Currently we don't use reverts, so we can ignore them, but they are here for when we do.
+        let timer = start_commit_timer("generate_plain_state_and_reverts");
+        let (plain_state, _reverts) =
+            bundle_state.into_plain_state_and_reverts(OriginalValuesKnown::Yes);
+        stop_timer(timer);
+
+        info!(
+            "Committing bundle state with {} accounts, {} contracts, {} storage changes",
+            plain_state.accounts.len(),
+            plain_state.contracts.len(),
+            plain_state.storage.len()
+        );
+
+        // Write Account State
+        let timer = start_commit_timer("account:committing_accounts_total");
+        self.commit_accounts(plain_state.accounts)?;
+        stop_timer(timer);
+
+        // Write Contract Code
+        let timer = start_commit_timer("contract:committing_contracts_total");
+        for (hash, bytecode) in plain_state.contracts {
```

Comment:
> nit: do you know if this includes only newly created contracts?

Yes this only returns newly created contracts.

> Also, do we ever delete the contracts from the db? It might not be big deal, but we are leaking memory if not. I would at least create an issue to track it, as we can't easily do it at the moment (because db key is their hash, and we don't know if there are multiple contracts with the same hash).

> Also, do we ever delete the contracts from the db

I think only if a contract is self destructed, but I don't think that is common

> It might not be big deal, but we are leaking memory if not. I would at least create an issue to track it, as we can't easily do it at the moment (because db key is their hash, and we don't know if there are multiple contracts with the same hash).

Because of the way REVM is designed, I don't think this is a problem, especially with self destruct being or about to be removed

Reply:
Yes, I agree it's probably too much work for little effort. I still think it's good to document the issue (e.g. creating github issue and adding link to it in the code).

Reply:
https://github.com/ethereum/trin/issues/1428 here is an issue for it. I don't think it will ever be fixed though as their is just so many more important things to improve, but I guess this is good for transparency so future people know this is a non-existent priority